rct229/rules/section21/section21rule5.py

```diff
@@ -0,0 +1,161 @@
+            # The heating_loop_conditioned_zone_area will include sum of all connected zones and indirectly zone areas.
+            heating_loop_conditioned_zone_area = ZERO.AREA
+            # The connected zones list, zones in this list can be residential, nonresidential, mixed or semi-heated
+            loop_zone_list = []
+
+            for fluid_loop in find_all("$.fluid_loops[*]", rmi_b):
+                # Make sure heating loop, and its heating is supplied by a boiler(s)
+                if (
+                    getattr_(fluid_loop, "fluid_loops", "type") == FLUID_LOOP.HEATING
+                    and fluid_loop["id"] in loop_attach_boiler_dict.keys()
+                ):
+                    boiler_loop_id = fluid_loop["id"]
+                    loop_zone_list.extend(
+                        loop_zone_list_w_area_dict[boiler_loop_id]["zone_list"]
+                    )
+                    heating_loop_conditioned_zone_area += loop_zone_list_w_area_dict[
+                        boiler_loop_id
+                    ]["total_area"]
+
+            # check indirectly conditioned zones, add them to the total area
+            for zone_id in zone_conditioning_category_dict.keys():
+                if (
+                    zone_conditioning_category_dict[zone_id]
+                    in [
+                        ZCC.CONDITIONED_MIXED,
+                        ZCC.CONDITIONED_RESIDENTIAL,
+                        ZCC.CONDITIONED_NON_RESIDENTIAL,
+                    ]
+                    and zone_id not in loop_zone_list
+                ):
+                    heating_loop_conditioned_zone_area += pint_sum(
+                        find_all(
+                            "$..floor_area",
+                            find_exactly_one_with_field_value(
+                                "$..zones[*]", "id", zone_id, rmi_b
+                            ),
+                        ),
+                        ZERO.AREA,
+                    )
+
+            num_boilers = len(find_all(".boilers[*]", rmi_b))
+            boiler_capacity_list = [
+                getattr_(boiler, "boiler", "rated_capacity")
+                for boiler in find_all(".boilers[*]", rmi_b)
+            ]
+
+            return {
+                "heating_loop_conditioned_zone_area": heating_loop_conditioned_zone_area,
+                "num_boilers": num_boilers,
+                "boiler_capacity_list": boiler_capacity_list,
+            }
+
+        def rule_check(self, context, calc_vals=None, data=None):
+            heating_loop_conditioned_zone_area = calc_vals[
+                "heating_loop_conditioned_zone_area"
+            ]
+            num_boilers = calc_vals["num_boilers"]
+            boiler_capacity_list = calc_vals["boiler_capacity_list"]
+            return {
+                heating_loop_conditioned_zone_area
+                <= HEATING_LOOP_CONDITIONED_AREA_THRESHOLD
+                and num_boilers == 1
+                or num_boilers == 2
```

Comment:
Some parentheses would be helpful to the reader if Black will allow them.

Reply:
Yep, that is allowed. Updated